core/src/state/world_state.rs

```diff
@@ -95,30 +90,178 @@ impl StateHandle {
     }
 }
 
+#[derive(Clone, Debug)]
+enum ListenerDefunctionalization {
+    RemoveLock {
+        lock: PlaneLockName,
+        uid: u64,
+        cluster: ClusterName,
+    },
+    Nothing,
+}
+
+#[derive(Debug)]
+struct Listener<T: Ord + Eq> {
+    target: T,
+    role: ListenerDefunctionalization,
+    sender: Sender<()>,
```

Comment:
The fact that every `Listener` has a `Sender` even if no receiver is held for it (e.g. in `apply` when handling a `LockMessage`) threw me off and makes it harder to grok the control flow.

The two usages of `Listener` are mostly disjoint and only have bit of behavior in common. One (for subscribing to sequence numbers) notifies async listeners and is arguably not part of state at all; the other (for subscribing to heartbeat times) must run sync and is part of the actual state update.

Instead of trying to generalize the union of that behavior, I would try to generalize the intersection of it.

My suggestion is, instead of `ListenerDefunctionalization` and `Listener`, implement a struct `MinQueue<C, T> where C: Ord + Eq` that implements a min-queue. Then, have a separate implementation for consuming each queue in `apply`.

Arguably, the external listener subscription should probably be moved out of `WorldState` eventually, but we will always want to check timestamp listeners in `apply`.